Feature: Implement public and private Trakt lists as sensors

downey-lv opened the pull request:
Created a new implementation of the highly requested Trakt lists. Tried to match your existing code as close as possible, but let me know if you want anything changed or have any suggestions for future contributions.

Currently doesn't support sorting,to keep the PR simpler, can implement that in the future.

Resolves #74, closes #75 

## Testing

### Private list
```
      - friendly_name: "Sonarr Watchlist"
        private_list: True
        list_id: "sonarr-watchlist"
```
![image](https://github.com/dylandoamaral/trakt-integration/assets/70920705/15b5ab66-d21c-41f6-ae42-adb2155a5c5c)

### Public list
```
      - friendly_name: "2024 Academy Awards"
        list_id: 26885014
        max_medias: 10
```

![image](https://github.com/dylandoamaral/trakt-integration/assets/70920705/e23b5eb9-7aae-4523-b7ea-744a3845e5d7)

### Filtered list
```
      - friendly_name: "Marvel Episodes"
        list_id: 1248149
        media_type: "episode"
```
![image](https://github.com/dylandoamaral/trakt-integration/assets/70920705/1e3cbabb-ee72-4a87-a645-3be94d57a2c5)

dylandoamaral commented:
I will also check it ASAP, thank you !

dylandoamaral commented:
It may be interesting to add the sorting strategy according to https://trakt.docs.apiary.io/#reference/users/list-items

downey-lv commented:
I started to look into it, but it was 1AM at that point 😃 Will check it out

dylandoamaral commented on custom_components/trakt_tv/schema.py:
```diff
@@ -41,6 +41,7 @@ def sensors_schema() -> Dict[str, Any]:
         "all_upcoming": upcoming_schema(),
         "next_to_watch": next_to_watch_schema(),
         "recommendation": recommendation_schema(),
+        "lists": All([lists_schema()]),
```
Not sure that you should add "All" here. Reading the documentation, I don't really know how to implement correctly though...

downey-lv replied:
`Schema([lists_schema()])` might be better here, will test it

dylandoamaral replied:
Yep would like to help more but voluptuous documentation is not clear

dylandoamaral commented on custom_components/trakt_tv/models/kind.py:
```diff
@@ -12,6 +12,13 @@ class CalendarInformation:
+@dataclass
+class ListInformation:
+    identifier: str
+    name: str
```
I prefere you to remove it and to handle the possible None value in sensors:
```

@property
    def name(self):
````

Otherwise, your sensor will be named "Trak List Name " instead of "Trak List Name"

dylandoamaral commented on custom_components/trakt_tv/sensor.py:
```diff
@@ -104,7 +123,12 @@ def name(self):
     def medias(self):
         if self.coordinator.data:
-            return self.coordinator.data.get(self.source, {}).get(self.trakt_kind, None)
+            medias = self.coordinator.data.get(self.source, {}).get(
```
I prefere:
```
    @property
    def medias(self):
        if not self.coordinator.data:
            return None

        if self.trakt_kind == TraktKind.LIST:
            try:
                name = self.config_entry["friendly_name"]
                return self.coordinator.data[self.source][self.trakt_kind][name]
            except KeyError:
                return None

        try:
            return self.coordinator.data[self.source][self.trakt_kind]
        except KeyError:
            return None
```
I find it clearer

dylandoamaral commented:
> I started to look into it, but it was 1AM at that point 😃 Will check it out

BTW the previous attempt added the sort feature if you need some inspiration:
https://github.com/dylandoamaral/trakt-integration/pull/75/files

Again thank you for your hard work !

downey-lv commented:
## Improvements
- Implemented your code review suggestions
- Implemented sorting
- Exposed trakt_rating to HA in case users prefer that over TMDB. Will look into populating this for all the other data sources in separate PR. Would probably allow to implement sorting everywhere as well.
- Moved all date formatting to a utils function

## Fixes
- Set trakt.request return type to dict as it's not returning ClientResponse

## Sorting demo
![image](https://github.com/dylandoamaral/trakt-integration/assets/70920705/e9006f11-bcf3-4571-ba89-166ffaecfc0f)

downey-lv commented:
I've synced with main and tested that everything works. Anything else before you can approve?

Razor5284 commented:
@dylandoamaral any chance of a merge on this please?

dylandoamaral commented:
I wait the PR to be merged with develop and then I will merge it with pleasure !

Razor5284 commented:
@downey-lv any chance of the above? :)

downey-lv commented:
Should be good to go now

Razor5284 commented:
@dylandoamaral Can we merge please? :)

Razor5284 commented:
@dylandoamaral 🙏 🙏

dylandoamaral commented:
I have a pending comment, waiting for the author response for 3 weeks. Tried on my PC but it is not working right now.

Razor5284 commented:
@downey-lv have you seen pending comment?

downey-lv commented:
Will try to take a look at what's not working soon, it's been a very busy time

dylandoamaral commented on custom_components/trakt_tv/models/media.py:
Change this to LF and then everything should be perfect :)

dylandoamaral commented:
About python, indeed you are right
